src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,144 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check, and to declare that a programmer has checked that
+/// these contracts have been upheld. They are not mutually exclusive, as can be
+/// seen in `unsafe fn` (more details on this can be found below).
+///
+/// Unsafe Rust add some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, some compiler
+/// intrinsics and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # The different meanings of `unsafe`
+///
+/// Depending on the place where it is used, the `unsafe` keyword meaning can
+/// change. See the following [discussion on Rust Internals] for more in-depth
+/// explanations.
+///
+/// Summarizing the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait Trait`: implementing this [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the invariants necessary to call the operations inside the
+/// block have been checked and are guaranteed to be valid.
```

Comment:
I don't think "invariant" is a great term here. You called it "contract" just two items above, which is better. Maybe use that consistently?

Reply:
I agree a unified vocabulary would be better, I'll fix it.